File: packages/lhc/test/fixtures/threads.ts
// Thread builders for Epic 02 (test plan §Test Substrate). All built through
// real intake; the multi-state and damaged-source builders then apply the
// sanctioned below-SDK writes for states the SDK cannot yet produce (drain
// and handlers land in Stories 1–3) — Story 0 proves read-back of what the
// builders wrote; Stories 2/4 confirm the states arise for real
// (coverage.md cross-story debt).
import { DatabaseSync } from "node:sqlite";
import {
  type DependencyGap,
  type Derivation,
  type DerivationMetadata,
  type DerivationState,
  intakeStream,
  type Lhc,
  type MessageEventInput,
  type SubjectKind,
  threads,
} from "../../src/index.js";
import { corruptTwoOpenTurns } from "./corrupt.js";
import { type TempStore, validEvent } from "./index.js";
import type { InferenceCallbacksDouble } from "./inference-callbacks-double.js";
import type { DerivationType } from "./model-call.js";

async function newThreadFile(store: TempStore): Promise<string> {
  const filePath = store.threadPath();
  const created = await threads.newThread({ filePath, registryPath: store.registryPath });
  if (!created.ok) throw new Error(`fixture thread creation failed: ${created.error.reason}`);
  return filePath;
}

async function send(filePath: string, batch: readonly MessageEventInput[]): Promise<void> {
  const result = await intakeStream.messageEvents({ filePath }, batch);
  if (!result.ok) throw new Error(`fixture batch failed: ${result.error.reason}`);
}

// n closed turns, each prompt + assistant text + explicit end. Turn ids are
// deterministic (t1..tn) per the turn store's id scheme.
export async function threadWithClosedTurns(
  store: TempStore,
  n: number,
): Promise<{ filePath: string; turnIds: string[] }> {
  const filePath = await newThreadFile(store);
  const turnIds: string[] = [];
  for (let i = 1; i <= n; i += 1) {
    await send(filePath, [
      validEvent("user_prompt", { payload: { text: `prompt for turn ${i}` } }),
      validEvent("assistant_text", { payload: { text: `answer for turn ${i}` } }),
      validEvent("turn_end"),
    ]);
    turnIds.push(`t${i}`);
  }
  return { filePath, turnIds };
}

// One closed turn containing a tool run: call + result pair by default;
// error and missing-result variants per the epic's outcome vocabulary.
export async function threadWithToolRun(
  store: TempStore,
  opts: { isError?: boolean; missingResult?: boolean; resultContent?: string } = {},
): Promise<{ filePath: string; turnId: string }> {
  const filePath = await newThreadFile(store);
  const batch: MessageEventInput[] = [
    validEvent("user_prompt", { payload: { text: "please run the tool" } }),
    validEvent("tool_call", {
      payload: { toolCallId: "call-fixture-1", toolName: "read_file", arguments: { path: "notes.txt" } },
    }),
  ];
  if (opts.missingResult !== true) {
    batch.push(
      validEvent("tool_result", {
        payload: {
          toolCallId: "call-fixture-1",
          content: opts.resultContent ?? "contents of notes.txt",
          isError: opts.isError ?? false,
        },
      }),
    );
  }
  batch.push(validEvent("turn_end"));
  await send(filePath, batch);
  return { filePath, turnId: "t1" };
}

// ── chunk read-back (Story 3): raw rows for boundary assertions ──

export interface ChunkSnapshot {
  chunks: Array<{
    chunkId: string;
    chunkOrder: number;
    status: "open" | "closed";
    accumulatedProjectedTokens: number;
  }>;
  members: Array<{ chunkId: string; turnId: string; memberIdx: number }>;
}

export function readChunks(filePath: string): ChunkSnapshot {
  const db = new DatabaseSync(filePath);
  try {
    const chunks = (
      db
        .prepare(
          `SELECT chunk_id, chunk_order, status, accumulated_projected_tokens
           FROM chunk ORDER BY chunk_order`,
        )
        .all() as unknown as Array<{
        chunk_id: string;
        chunk_order: number | bigint;
        status: string;
        accumulated_projected_tokens: number | bigint;
      }>
    ).map((row) => ({
      chunkId: row.chunk_id,
      chunkOrder: Number(row.chunk_order),
      status: row.status as "open" | "closed",
      accumulatedProjectedTokens: Number(row.accumulated_projected_tokens),
    }));
    const members = (
      db
        .prepare(
          `SELECT cm.chunk_id, cm.turn_id, cm.member_idx FROM chunk_member cm
           JOIN chunk c ON c.chunk_id = cm.chunk_id
           ORDER BY c.chunk_order, cm.member_idx`,
        )
        .all() as unknown as Array<{
        chunk_id: string;
        turn_id: string;
        member_idx: number | bigint;
      }>
    ).map((row) => ({
      chunkId: row.chunk_id,
      turnId: row.turn_id,
      memberIdx: Number(row.member_idx),
    }));
    return { chunks, members };
  } finally {
    db.close();
  }
}

// ── derivation read-back and the sanctioned state writer ───────

interface RawFormRow {
  subject_kind: string;
  subject_id: string;
  derivation_type: string;
  state: string;
  content: string | null;
  reason: string | null;
  metadata: string | null;
  source_version: number | bigint;
  gaps: string | null;
  derived_at: string | null;
}

export function readDerivedForms(filePath: string): Derivation[] {
  const db = new DatabaseSync(filePath);
  try {
    const rows = db
      .prepare(
        `SELECT subject_kind, subject_id, derivation_type, state, content, reason, metadata,
                source_version, gaps, derived_at
         FROM derivation ORDER BY subject_kind, subject_id, derivation_type`,
      )
      .all() as unknown as RawFormRow[];
    return rows.map((row) => {
      const record: Derivation = {
        subjectKind: row.subject_kind as SubjectKind,
        subjectId: row.subject_id,
        derivationType: row.derivation_type as DerivationType,
        state: row.state as DerivationState,
        sourceVersion: Number(row.source_version),
      };
      if (row.content !== null) record.content = row.content;
      if (row.reason !== null) record.reason = row.reason;
      if (row.metadata !== null) {
        record.metadata = JSON.parse(row.metadata) as DerivationMetadata;
      }
      if (row.gaps !== null) record.gaps = JSON.parse(row.gaps) as DependencyGap[];
      if (row.derived_at !== null) record.derivedAt = row.derived_at;
      return record;
    });
  } finally {
    db.close();
  }
}

// Below-SDK state writer, sanctioned like corrupt.ts: ready/failed/blocked
// are unreachable through the SDK until the drain (Story 1) and handlers
// (Stories 2–3) land. UPDATE-only on purpose — the pending row exists from
// enqueue, mirroring the production completion contract.
export function setFormState(
  filePath: string,
  target: { subjectKind: SubjectKind; subjectId: string; derivationType: DerivationType },
  update: {
    state: DerivationState;
    content?: string;
    reason?: string;
    metadata?: DerivationMetadata;
    derivedAt?: string;
  },
): void {
  const db = new DatabaseSync(filePath);
  try {
    const changed = db
      .prepare(
        `UPDATE derivation
         SET state = ?, content = ?, reason = ?, metadata = ?, derived_at = ?
         WHERE subject_kind = ? AND subject_id = ? AND derivation_type = ?`,
      )
      .run(
        update.state,
        update.content ?? null,
        update.reason ?? null,
        update.metadata === undefined ? null : JSON.stringify(update.metadata),
        update.derivedAt ?? null,
        target.subjectKind,
        target.subjectId,
        target.derivationType,
      );
    if (Number(changed.changes) !== 1) {
      throw new Error(
        `fixture setFormState hit ${String(changed.changes)} rows for ${target.subjectKind}/${target.subjectId}/${target.derivationType}; expected the pending row from enqueue`,
      );
    }
  } finally {
    db.close();
  }
}

// ── multi-state thread: every form state in one file ─────────────

export interface MultiStateClaim {
  subjectKind: SubjectKind;
  subjectId: string;
  derivationType: DerivationType;
  state: DerivationState;
}

export async function multiStateThread(store: TempStore): Promise<{ filePath: string; expected: MultiStateClaim[] }> {
  const filePath = await newThreadFile(store);
  // Turn 1: prompt + tool run; turn 2: prompt. Intake's enqueue leaves every
  // form pending; the writes below move a subset into the other three states.
  await send(filePath, [
    validEvent("user_prompt", { payload: { text: "first prompt" } }),
    validEvent("assistant_text", { payload: { text: "working on it" } }),
    validEvent("tool_call", {
      payload: { toolCallId: "call-ms-1", toolName: "read_file", arguments: { path: "a.txt" } },
    }),
    validEvent("tool_result", {
      payload: { toolCallId: "call-ms-1", content: "contents of a.txt", isError: false },
    }),
    validEvent("turn_end"),
    validEvent("user_prompt", { payload: { text: "second prompt" } }),
    validEvent("turn_end"),
  ]);
  const derivedAt = "2026-06-10T12:00:00.000Z";
  setFormState(
    filePath,
    { subjectKind: "message", subjectId: "m1", derivationType: "smoothed_prompt" },
    { state: "ready", content: "smoothed(fixture:first prompt)", derivedAt },
  );
  // Tool-activity outcome lives in machine-readable metadata, never inside
  // model-authored content (FC-0.3).
  setFormState(
    filePath,
    { subjectKind: "message", subjectId: "m4", derivationType: "tool_result_summary" },
    {
      state: "ready",
      content: "toolresult(fixture:contents of a.txt)",
      metadata: { outcome: "succeeded" },
      derivedAt,
    },
  );
  setFormState(
    filePath,
    { subjectKind: "turn", subjectId: "t1", derivationType: "turn_rendering" },
    { state: "failed", reason: "provider_failure: scripted exhaustion (fixture)" },
  );
  setFormState(
    filePath,
    { subjectKind: "turn", subjectId: "t2", derivationType: "turn_rendering" },
    { state: "blocked", reason: "source_damaged: manufactured damage (fixture)" },
  );
  return {
    filePath,
    expected: [
      { subjectKind: "message", subjectId: "m1", derivationType: "smoothed_prompt", state: "ready" },
      { subjectKind: "message", subjectId: "m4", derivationType: "tool_result_summary", state: "ready" },
      { subjectKind: "message", subjectId: "m6", derivationType: "smoothed_prompt", state: "pending" },
      { subjectKind: "turn", subjectId: "t1", derivationType: "turn_rendering", state: "failed" },
      { subjectKind: "turn", subjectId: "t1", derivationType: "smooth_turn_compression", state: "pending" },
      { subjectKind: "turn", subjectId: "t2", derivationType: "turn_rendering", state: "blocked" },
      { subjectKind: "turn", subjectId: "t2", derivationType: "smooth_turn_compression", state: "pending" },
    ],
  };
}

// ── damaged-source thread (Epic 01 corruption definition) ────────

// A closed turn with its turn_derivation work queued, then the Epic 01
// two-open-turns corruption — the damage a turn handler must find and the
// intake pipeline refuses (turn_state_corrupt).
export async function damagedSourceThread(store: TempStore): Promise<{ filePath: string; turnId: string }> {
  const { filePath, turnIds } = await threadWithClosedTurns(store, 1);
  // Open a second turn through real intake; the corruption writer then adds
  // another open row — two open turns, the Epic 01 invariant violation.
  await send(filePath, [validEvent("user_prompt", { payload: { text: "left open" } })]);
  corruptTwoOpenTurns(filePath);
  const turnId = turnIds[0];
  if (turnId === undefined) throw new Error("fixture invariant: one closed turn expected");
  return { filePath, turnId };
}

// ── fallback-rendering thread (TC-3.2's state, shared) ─────────────

// TC-3.2's fallback-rendering state as one shared builder (coverage.md
// cross-story debt: TC-4.4 consumes this exact scenario; don't rebuild it by
// hand). Expects a manual-mode SDK with retry budget 3 and zero backoff:
// scripts the double to exhaust the prompt's smoothing, then drains — the
// smoothing lands failed (its work row deleted at exhaustion, DD-1) and the
// turn rendering lands ready from the raw floor.
export const GAPPED_SMOOTHING_REASON = "provider_failure: scripted smoothing exhaustion";

export async function gappedRenderingThread(
  store: TempStore,
  sdk: Lhc,
  double: InferenceCallbacksDouble,
): Promise<{ filePath: string; messageId: string; turnId: string }> {
  const filePath = await newThreadFile(store);
  double.failKind("prompt_smoothing", 3, {
    retryable: true,
    reason: GAPPED_SMOOTHING_REASON,
  });
  await send(filePath, [
    validEvent("user_prompt", { payload: { text: "gapped prompt" } }),
    validEvent("assistant_text", { payload: { text: "gapped answer" } }),
    validEvent("turn_end"),
  ]);
  const drained = await sdk.work.drain({ filePath });
  if (!drained.ok) throw new Error(`fixture drain failed: ${drained.error.reason}`);
  setFormState(
    filePath,
    { subjectKind: "message", subjectId: "m1", derivationType: "smoothed_prompt" },
    { state: "failed", reason: GAPPED_SMOOTHING_REASON },
  );
  const forms = readDerivedForms(filePath);
  const smoothing = forms.find((f) => f.subjectId === "m1" && f.derivationType === "smoothed_prompt");
  const rendering = forms.find((f) => f.subjectId === "t1" && f.derivationType === "turn_rendering");
  if (smoothing?.state !== "failed" || rendering?.state !== "ready" || rendering.gaps !== undefined) {
    throw new Error("fixture invariant: failed smoothing under a ready fallback rendering expected");
  }
  return { filePath, messageId: "m1", turnId: "t1" };
}

File: packages/lhc-py/tests/fixtures/threads.py
"""Ported from packages/lhc/test/fixtures/threads.ts. Phase 1 skeleton.

Thread builders for Epic 02 (test plan §Test Substrate). Types/constants are
real; SDK-calling and below-SDK DB helpers are skeletons.
"""

from __future__ import annotations

import json
from dataclasses import asdict, is_dataclass
from typing import TYPE_CHECKING, Literal, NotRequired, TypedDict

from lhc import intake_stream, threads
from lhc.intake_stream import MessageEventInput
from lhc.sdk import Lhc
from lhc.shared_tech.derivation import (
    DependencyGap,
    Derivation,
    DerivationMetadata,
    DerivationState,
    ProviderProvenance,
    SubjectKind,
)
from lhc.shared_tech.storage import open_database
from .corrupt import corrupt_two_open_turns
from .inference_callbacks_double import InferenceCallbacksDouble
from .model_call import DerivationType

if TYPE_CHECKING:
    from . import TempStore


class ChunkRow(TypedDict):
    chunkId: str
    chunkOrder: int
    status: Literal["open", "closed"]
    accumulatedProjectedTokens: int


class ChunkMemberRow(TypedDict):
    chunkId: str
    turnId: str
    memberIdx: int


class ChunkSnapshot(TypedDict):
    chunks: list[ChunkRow]
    members: list[ChunkMemberRow]


class FormStateTarget(TypedDict):
    subjectKind: SubjectKind
    subjectId: str
    derivationType: DerivationType


class FormStateUpdate(TypedDict):
    state: DerivationState
    content: NotRequired[str]
    reason: NotRequired[str]
    metadata: NotRequired[DerivationMetadata]
    derivedAt: NotRequired[str]


class MultiStateClaim(TypedDict):
    subjectKind: SubjectKind
    subjectId: str
    derivationType: DerivationType
    state: DerivationState


class ThreadWithClosedTurnsResult(TypedDict):
    filePath: str
    turnIds: list[str]


class ThreadWithToolRunResult(TypedDict):
    filePath: str
    turnId: str


class MultiStateThreadResult(TypedDict):
    filePath: str
    expected: list[MultiStateClaim]


class DamagedSourceThreadResult(TypedDict):
    filePath: str
    turnId: str


class GappedRenderingThreadResult(TypedDict):
    filePath: str
    messageId: str
    turnId: str


class ToolRunOpts(TypedDict, total=False):
    isError: bool
    missingResult: bool
    resultContent: str


# TC-3.2's fallback-rendering state as one shared builder (coverage.md
# cross-story debt: TC-4.4 consumes this exact scenario; don't rebuild it by
# hand). Expects a manual-mode SDK:
# scripts the prompt's smoothing to fail, then drains — the smoothing lands
# failed (its work row deleted) and the
# turn rendering lands ready from the raw floor.
GAPPED_SMOOTHING_REASON = "provider_failure: scripted smoothing failure"


async def _new_thread_file(store: TempStore) -> str:
    file_path = store.thread_path()
    created = await threads.new_thread(
        {"filePath": file_path, "registryPath": store.registry_path}
    )
    if not created.ok:
        raise RuntimeError(f"fixture thread creation failed: {created.error.reason}")
    return file_path


async def _send(file_path: str, batch: list[MessageEventInput] | tuple[MessageEventInput, ...]) -> None:
    result = await intake_stream.message_events({"filePath": file_path}, batch)
    if not result.ok:
        raise RuntimeError(f"fixture batch failed: {result.error.reason}")


async def thread_with_closed_turns(
    store: TempStore,
    n: int,
) -> ThreadWithClosedTurnsResult:
    from . import valid_event

    file_path = await _new_thread_file(store)
    turn_ids: list[str] = []
    for i in range(1, n + 1):
        await _send(
            file_path,
            [
                valid_event(
                    "user_prompt",
                    {"payload": {"text": f"prompt for turn {i}"}},
                ),
                valid_event(
                    "assistant_text",
                    {"payload": {"text": f"answer for turn {i}"}},
                ),
                valid_event("turn_end"),
            ],
        )
        turn_ids.append(f"t{i}")
    return {"filePath": file_path, "turnIds": turn_ids}


async def thread_with_tool_run(
    store: TempStore,
    opts: ToolRunOpts | None = None,
) -> ThreadWithToolRunResult:
    from . import valid_event

    options = opts if opts is not None else {}
    file_path = await _new_thread_file(store)
    batch: list[MessageEventInput] = [
        valid_event(
            "user_prompt",
            {"payload": {"text": "please run the tool"}},
        ),
        valid_event(
            "tool_call",
            {
                "payload": {
                    "toolCallId": "call-fixture-1",
                    "toolName": "read_file",
                    "arguments": {"path": "notes.txt"},
                }
            },
        ),
    ]
    if options.get("missingResult") is not True:
        batch.append(
            valid_event(
                "tool_result",
                {
                    "payload": {
                        "toolCallId": "call-fixture-1",
                        "content": options.get(
                            "resultContent",
                            "contents of notes.txt",
                        ),
                        "isError": options.get("isError", False),
                    }
                },
            )
        )
    batch.append(valid_event("turn_end"))
    await _send(file_path, batch)
    return {"filePath": file_path, "turnId": "t1"}


# ── chunk read-back (Story 3): raw rows for boundary assertions ──


def read_chunks(file_path: str) -> ChunkSnapshot:
    db = open_database(file_path)
    try:
        chunk_rows = db.prepare(
            """SELECT chunk_id, chunk_order, status, accumulated_projected_tokens
               FROM chunk ORDER BY chunk_order"""
        ).all()
        chunks: list[ChunkRow] = [
            {
                "chunkId": str(row["chunk_id"]),
                "chunkOrder": int(row["chunk_order"]),
                "status": row["status"],
                "accumulatedProjectedTokens": int(
                    row["accumulated_projected_tokens"]
                ),
            }
            for row in chunk_rows
        ]
        member_rows = db.prepare(
            """SELECT cm.chunk_id, cm.turn_id, cm.member_idx
               FROM chunk_member cm
               JOIN chunk c ON c.chunk_id = cm.chunk_id
               ORDER BY c.chunk_order, cm.member_idx"""
        ).all()
        members: list[ChunkMemberRow] = [
            {
                "chunkId": str(row["chunk_id"]),
                "turnId": str(row["turn_id"]),
                "memberIdx": int(row["member_idx"]),
            }
            for row in member_rows
        ]
        return {"chunks": chunks, "members": members}
    finally:
        db.close()


# ── derivation read-back and the sanctioned state writer ───────


def read_derived_forms(file_path: str) -> list[Derivation]:
    db = open_database(file_path)
    try:
        rows = db.prepare(
            """SELECT subject_kind, subject_id, derivation_type, state, content,
                      reason, metadata, source_version, gaps, derived_at
               FROM derivation
               ORDER BY subject_kind, subject_id, derivation_type"""
        ).all()
        records: list[Derivation] = []
        for row in rows:
            metadata = None
            if row["metadata"] is not None:
                raw_metadata = json.loads(str(row["metadata"]))
                # Drift guard (TS surfaces parsed JSON verbatim, so unknown
                # keys fail deep-equality there; the dataclass rebuild here
                # would silently drop them without this check).
                _known = {
                    "outcome", "lastError", "discardReason", "fallbackFloor",
                    "fallbackUsed", "inferenceAttempted", "inferenceSucceeded",
                    "sizeDisposition", "provenance",
                }
                _extra = set(raw_metadata) - _known
                if _extra:
                    raise AssertionError(
                        f"derivation metadata carries undeclared keys: {sorted(_extra)}"
                    )
                provenance = raw_metadata.get("provenance")
                if isinstance(provenance, dict):
                    _pextra = set(provenance) - {"provider", "model", "prompt"}
                    if _pextra:
                        raise AssertionError(
                            f"provenance carries undeclared keys: {sorted(_pextra)}"
                        )
                metadata = DerivationMetadata(
                    outcome=raw_metadata.get("outcome"),
                    last_error=raw_metadata.get("lastError"),
                    discard_reason=raw_metadata.get("discardReason"),
                    fallback_floor=raw_metadata.get("fallbackFloor"),
                    fallback_used=raw_metadata.get("fallbackUsed"),
                    inference_attempted=raw_metadata.get("inferenceAttempted"),
                    inference_succeeded=raw_metadata.get("inferenceSucceeded"),
                    size_disposition=raw_metadata.get("sizeDisposition"),
                    provenance=(
                        ProviderProvenance(
                            provider=provenance["provider"],
                            model=provenance["model"],
                            prompt=provenance["prompt"],
                        )
                        if isinstance(provenance, dict)
                        else None
                    ),
                )
            gaps = None
            if row["gaps"] is not None:
                gaps = [
                    DependencyGap(
                        subject_kind=gap["subjectKind"],
                        subject_id=gap["subjectId"],
                        derivation_type=gap["derivationType"],
                    )
                    for gap in json.loads(str(row["gaps"]))
                ]
            records.append(
                Derivation(
                    subject_kind=row["subject_kind"],
                    subject_id=str(row["subject_id"]),
                    derivation_type=str(row["derivation_type"]),
                    state=row["state"],
                    source_version=int(row["source_version"]),
                    content=(
                        str(row["content"]) if row["content"] is not None else None
                    ),
                    reason=(
                        str(row["reason"]) if row["reason"] is not None else None
                    ),
                    metadata=metadata,
                    gaps=gaps,
                    derived_at=(
                        str(row["derived_at"])
                        if row["derived_at"] is not None
                        else None
                    ),
                )
            )
        return records
    finally:
        db.close()


def _metadata_json(value: DerivationMetadata | dict[str, object]) -> str:
    if isinstance(value, dict):
        data = value
    elif is_dataclass(value):
        snake = asdict(value)
        key_map = {
            "last_error": "lastError",
            "discard_reason": "discardReason",
            "fallback_floor": "fallbackFloor",
            "fallback_used": "fallbackUsed",
            "inference_attempted": "inferenceAttempted",
            "inference_succeeded": "inferenceSucceeded",
            "size_disposition": "sizeDisposition",
        }
        data = {
            key_map.get(key, key): item
            for key, item in snake.items()
            if item is not None
        }
    else:
        data = value
    return json.dumps(data, separators=(",", ":"), ensure_ascii=False)


# Below-SDK state writer, sanctioned like corrupt.ts: ready/failed/blocked
# are unreachable through the SDK until the drain (Story 1) and handlers
# (Stories 2–3) land. UPDATE-only on purpose — the pending row exists from
# enqueue, mirroring the production completion contract.
def set_form_state(
    file_path: str,
    target: FormStateTarget,
    update: FormStateUpdate,
) -> None:
    db = open_database(file_path)
    try:
        metadata = update.get("metadata")
        changed = db.prepare(
            """UPDATE derivation
               SET state = ?, content = ?, reason = ?, metadata = ?, derived_at = ?
               WHERE subject_kind = ? AND subject_id = ? AND derivation_type = ?"""
        ).run(
            update["state"],
            update.get("content"),
            update.get("reason"),
            _metadata_json(metadata) if metadata is not None else None,
            update.get("derivedAt"),
            target["subjectKind"],
            target["subjectId"],
            target["derivationType"],
        )
        if int(changed.changes) != 1:
            raise RuntimeError(
                "fixture setFormState hit "
                f"{changed.changes} rows for "
                f'{target["subjectKind"]}/{target["subjectId"]}/{target["derivationType"]}; '
                "expected the pending row from enqueue"
            )
    finally:
        db.close()


# ── multi-state thread: every form state in one file ─────────────


async def multi_state_thread(store: TempStore) -> MultiStateThreadResult:
    from . import valid_event

    file_path = await _new_thread_file(store)
    await _send(
        file_path,
        [
            valid_event("user_prompt", {"payload": {"text": "first prompt"}}),
            valid_event(
                "assistant_text",
                {"payload": {"text": "working on it"}},
            ),
            valid_event(
                "tool_call",
                {
                    "payload": {
                        "toolCallId": "call-ms-1",
                        "toolName": "read_file",
                        "arguments": {"path": "a.txt"},
                    }
                },
            ),
            valid_event(
                "tool_result",
                {
                    "payload": {
                        "toolCallId": "call-ms-1",
                        "content": "contents of a.txt",
                        "isError": False,
                    }
                },
            ),
            valid_event("turn_end"),
            valid_event("user_prompt", {"payload": {"text": "second prompt"}}),
            valid_event("turn_end"),
        ],
    )
    derived_at = "2026-06-10T12:00:00.000Z"
    set_form_state(
        file_path,
        {
            "subjectKind": "message",
            "subjectId": "m1",
            "derivationType": "smoothed_prompt",
        },
        {
            "state": "ready",
            "content": "smoothed(fixture:first prompt)",
            "derivedAt": derived_at,
        },
    )
    set_form_state(
        file_path,
        {
            "subjectKind": "message",
            "subjectId": "m4",
            "derivationType": "tool_result_summary",
        },
        {
            "state": "ready",
            "content": "toolresult(fixture:contents of a.txt)",
            "metadata": {"outcome": "succeeded"},
            "derivedAt": derived_at,
        },
    )
    set_form_state(
        file_path,
        {
            "subjectKind": "turn",
            "subjectId": "t1",
            "derivationType": "turn_rendering",
        },
        {"state": "failed", "reason": "provider_failure: scripted failure (fixture)"},
    )
    set_form_state(
        file_path,
        {
            "subjectKind": "turn",
            "subjectId": "t2",
            "derivationType": "turn_rendering",
        },
        {
            "state": "blocked",
            "reason": "source_damaged: manufactured damage (fixture)",
        },
    )
    expected: list[MultiStateClaim] = [
        {"subjectKind": "message", "subjectId": "m1", "derivationType": "smoothed_prompt", "state": "ready"},
        {"subjectKind": "message", "subjectId": "m4", "derivationType": "tool_result_summary", "state": "ready"},
        {"subjectKind": "message", "subjectId": "m6", "derivationType": "smoothed_prompt", "state": "pending"},
        {"subjectKind": "turn", "subjectId": "t1", "derivationType": "turn_rendering", "state": "failed"},
        {"subjectKind": "turn", "subjectId": "t1", "derivationType": "pre_detailed_assembly", "state": "pending"},
        {"subjectKind": "turn", "subjectId": "t2", "derivationType": "turn_rendering", "state": "blocked"},
        {"subjectKind": "turn", "subjectId": "t2", "derivationType": "pre_detailed_assembly", "state": "pending"},
    ]
    return {"filePath": file_path, "expected": expected}


# ── damaged-source thread (Epic 01 corruption definition) ────────


async def damaged_source_thread(store: TempStore) -> DamagedSourceThreadResult:
    from . import valid_event

    built = await thread_with_closed_turns(store, 1)
    file_path = built["filePath"]
    await _send(
        file_path,
        [valid_event("user_prompt", {"payload": {"text": "left open"}})],
    )
    corrupt_two_open_turns(file_path)
    if not built["turnIds"]:
        raise RuntimeError("fixture invariant: one closed turn expected")
    return {"filePath": file_path, "turnId": built["turnIds"][0]}


# ── fallback-rendering thread (TC-3.2's state, shared) ─────────────


async def gapped_rendering_thread(
    store: TempStore,
    sdk: Lhc,
    double: InferenceCallbacksDouble,
) -> GappedRenderingThreadResult:
    from . import valid_event

    file_path = await _new_thread_file(store)
    double.fail_kind(
        "prompt_smoothing",
        1,
        {"reason": GAPPED_SMOOTHING_REASON},
    )
    await _send(
        file_path,
        [
            valid_event("user_prompt", {"payload": {"text": "gapped prompt"}}),
            valid_event(
                "assistant_text",
                {"payload": {"text": "gapped answer"}},
            ),
            valid_event("turn_end"),
        ],
    )
    drained = await sdk.work.drain({"filePath": file_path})
    if not drained.ok:
        raise RuntimeError(f"fixture drain failed: {drained.error.reason}")
    set_form_state(
        file_path,
        {
            "subjectKind": "message",
            "subjectId": "m1",
            "derivationType": "smoothed_prompt",
        },
        {"state": "failed", "reason": GAPPED_SMOOTHING_REASON},
    )
    forms = read_derived_forms(file_path)
    smoothing = next(
        (
            form
            for form in forms
            if form.subject_id == "m1"
            and form.derivation_type == "smoothed_prompt"
        ),
        None,
    )
    rendering = next(
        (
            form
            for form in forms
            if form.subject_id == "t1"
            and form.derivation_type == "turn_rendering"
        ),
        None,
    )
    if (
        smoothing is None
        or smoothing.state != "failed"
        or rendering is None
        or rendering.state != "ready"
        or rendering.gaps is not None
    ):
        raise RuntimeError(
            "fixture invariant: failed smoothing under a ready fallback rendering expected"
        )
    return {"filePath": file_path, "messageId": "m1", "turnId": "t1"}


__all__ = [
    "ChunkSnapshot",
    "DamagedSourceThreadResult",
    "GAPPED_SMOOTHING_REASON",
    "GappedRenderingThreadResult",
    "MultiStateClaim",
    "MultiStateThreadResult",
    "ThreadWithClosedTurnsResult",
    "ThreadWithToolRunResult",
    "damaged_source_thread",
    "gapped_rendering_thread",
    "multi_state_thread",
    "read_chunks",
    "read_derived_forms",
    "set_form_state",
    "thread_with_closed_turns",
    "thread_with_tool_run",
]
